server/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java

```diff
@@ -92,7 +97,7 @@ public Explanation explain(int topLevelDocId, IndexSearcher searcher, RescoreCon
 
         // NOTE: we don't use Lucene's Rescorer.explain because we want to insert our own description with which ScoreMode was used.  Maybe
         // we should add QueryRescorer.explainCombine to Lucene?
-        if (rescoreExplain != null && rescoreExplain.isMatch()) {
+        if (rescoreContext.isRescored(topLevelDocId) && rescoreExplain != null && rescoreExplain.isMatch()) {
```

Comment:
We always build the rescore explanation even when `rescoreContext.isRescored(topLevelDocId)` is false. We could avoid this by building the primary explanation first and return it directly if `rescoreContext.isRescored(topLevelDocId)` is false ?